### PR TITLE
ENH: Initial implementation of Rotation.match_vectors

### PR DESCRIPTION
**Pull request by adbugger**
Solution of Wahba's problem using ~Davenport's q-method~ Markley's SVD method.

### CODE REVIEW

File: scipy/spatial/transform/rotation.py [hunk 1]
```diff
@@ -866,3 +868,94 @@ def __getitem__(self, indexer):
                   array.
         """
         return self.__class__(self._quat[indexer], normalized=True)
+
+    @classmethod
+    def estimate(cls, vectors, transformed_vectors, coeff=None):
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: There are potentially many ways to estimate a rotation, I suggest to name it as `match_vectors` perhaps.

Definitely not `coef`, name it `weights`.

File: scipy/spatial/transform/rotation.py [hunk 2]
```diff
@@ -866,3 +868,94 @@ def __getitem__(self, indexer):
...
+
+        Given `vectors` expressed in an initial frame A, corresponding
+        `transformed_vectors` expressed in frame B, estimate the rotation that
+        transforms frame A to frame B. Also known as Wahba's problem.
+
+        The rotation is estimated using Davenport's q-method [1]_.
+
+        Parameters
+        ----------
+        vectors : array_like, shape (N, 3) or (3,)
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: You can't estimate rotation using a single vector.

Also remember we discussed what to do with vector norms. I think we should introduce `normalized=False` for that.

File: scipy/spatial/transform/rotation.py [hunk 3]
```diff
@@ -866,3 +868,94 @@ def __getitem__(self, indexer):
...
+            Vectors expressed in initial frame A.
+        transformed_vectors : array_like, shape (N, 3) or (3,)
+            Result of applying an arbitrary rotation to `vectors`. Shape must
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: Use the similar description.

Also you can/should emphasize that two sets of vectors are typically measured somehow in the two frames. For example, directions to certain stars are known from datum in the global frame and they are measured (observed) in spacecraft frame, but with some angular errors. This way the attitude of spacecraft is determined. This is how conceptually this algorithm is used.

So your description is conceptually misleading, while formally sort of correct.

File: scipy/spatial/transform/rotation.py [hunk 4]
```diff
@@ -866,3 +868,94 @@ def __getitem__(self, indexer):
...
+
+    @classmethod
+    def estimate(cls, vectors, transformed_vectors, coeff=None):
+        # TODO: Better description of coeff?
+        """Estimate the rotation that best transforms a set of vectors to
+        another.
+
+        Given `vectors` expressed in an initial frame A, corresponding
+        `transformed_vectors` expressed in frame B, estimate the rotation that
+        transforms frame A to frame B. Also known as Wahba's problem.
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: So the estimated DCM C does `v_A = C v_B`?

Also explain that it is estimated in (weighted) least-squares sense, writing objective function would be appropriate.

File: scipy/spatial/transform/rotation.py [hunk 5]
```diff
@@ -866,3 +868,94 @@ def __getitem__(self, indexer):
...
+
+        The rotation is estimated using Davenport's q-method [1]_.
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: I suggest to use Markley SVD method, it is cleaner to implement and additionally conveniently provides the covariance matrix for the estimation.

And I suggest to return the covariance to be slightly more advanced and it is useful sometimes. Considering that, I think you should make a standalone function like `match_ vectors`.

What do you think of this suggestion?

File: scipy/spatial/transform/tests/test_rotation.py [hunk 6]
```diff
@@ -778,3 +778,11 @@ def test_quat_ownership():
 
     r._quat[0] = np.array([0, -1, 0, 0])
     assert_allclose(s._quat[0], np.array([1, 0, 0, 0]))
+
+
+def test_estimate_no_rotation():
```
Comment (nmayorov) [inline scipy/spatial/transform/tests/test_rotation.py]: Add a test with non-trivial rotation, two vectors and many vectors, without noise and with noise. Obviously with noise you won't get the rotation back precisely, but you can control the level of error, especially if you compute the covariance matrix.

**Review by nmayorov — COMMENTED** [review]
Made some comments (from the phone).

File: scipy/spatial/transform/rotation.py (hunk 4 above)
Comment (adbugger) [reply scipy/spatial/transform/rotation.py]: No, I meant `v_B = C v_A`. I will clear this confusion in the next push.

File: scipy/spatial/transform/rotation.py (hunk 5 above)
Comment (adbugger) [reply scipy/spatial/transform/rotation.py]: When you say covariance matrix, do you mean the singular value decomposition of `A_opt` or the error covariance matrix? [Paper](https://pdfs.semanticscholar.org/f151/97510b1e501876a6f6d8683ac69ab1ef8d39.pdf) for reference.

I'm not exactly sure about the need for a standalone function, but the Markley SVD method does seem cleaner and more informative.

**Comment by nmayorov** [comment]
I mean small vector errors covariance.

Well, yes I believe standalone function is not necessary.
Sent using Zoho Mail


---- On Sun, 24 Jun 2018 22:14:03 +0500 notifications@github.com wrote ----

@adbugger commented on this pull request.

In scipy/spatial/transform/rotation.py:

> @@ -866,3 +868,94 @@ def __getitem__(self, indexer):
                   array.
         """
         return self.__class__(self._quat[indexer], normalized=True)
+
+    @classmethod
+    def estimate(cls, vectors, transformed_vectors, coeff=None):
+        # TODO: Better description of coeff?
+        """Estimate the rotation that best transforms a set of vectors to
+        another.
+
+        Given `vectors` expressed in an initial frame A, corresponding
+        `transformed_vectors` expressed in frame B, estimate the rotation that
+        transforms frame A to frame B. Also known as Wahba's problem.
+
+        The rotation is estimated using Davenport's q-method [1]_.
When you say covariance matrix, do you mean the singular value decomposition of A_opt or the error covariance matrix? Paper for reference.

I'm not exactly sure about the need for a standalone function, but the Markley SVD method does seem cleaner and more informative.

File: scipy/spatial/transform/rotation.py (hunk 4 above)
Comment (nmayorov) [reply scipy/spatial/transform/rotation.py]: I think I would prefer having `v_A, v_B` and `C^A_B`, such that `v_A = C^A_B v_B`.

File: scipy/spatial/transform/rotation.py (hunk 5 above)
Comment (nmayorov) [reply scipy/spatial/transform/rotation.py]: I mean covariance of attitude error as a small rotation vector. 

Well, I think you are right that we don't need a separate function.

Comment (nmayorov) [reply scipy/spatial/transform/rotation.py]: To clarify, I think you don't need to consider the scale of the covariance only its "structure".

Check the formula (52) in the paper I sent you, what I suggest to return is  the matrix part withou "lambda_0 sigma_tot^2". At the moment though I forgot about fine points of this derivation, you need to figure it out. My idea is that we should return something like "sensitivity matrix" and the covariance can be computed by scaling it appropriately.

**Comment by adbugger** [comment]
I have updated the implementation. Now we return an appropriate 'sensitivity_matrix`, along with the estimated rotation.

I need some help with the testing part here. I know I can create test cases by starting with fixed `vectors` and a given rotation, applying the rotation and adding [gaussian?] noise to get `transformed_vectors`. How do I ensure that the estimated rotation is within the error bounds? For example, if we were returning the full covariance matrix, I could possibly compute the error vector between `estimated_rotation.apply(vectors)` and `transformed_vectors`, and check that the result was within the tolerance given by the covariance matrix. Now since our `sensitivity_matrix` is scaled, I'm not sure how to proceed.

File: scipy/spatial/transform/rotation.py [hunk 7]
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
...
+        This is also known as Wahba's problem and can be formulated as a
+        minimization problem with the following loss function corresponging to
+        a rotation matrix :math:`A`:
+
+        .. math::
+
+            L(A) = \\frac{1}{2} \\sum_{i = 1}^{n} a_i \\lVert \\mathbf{b}_i -
+            A \\mathbf{r}_i \\rVert^2 ,
+
+        where :math:`a_i`'s are the `weights` corresponding to each vector, and
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: w_i for weights?

File: scipy/spatial/transform/rotation.py [hunk 8]
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
                   array.
         """
         return self.__class__(self._quat[indexer], normalized=True)
+
+    @classmethod
+    def match_vectors(cls, vectors, transformed_vectors, weights=None):
+        # TODO: Better description of coeff?
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: Remove.

File: scipy/spatial/transform/rotation.py [hunk 9]
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
...
+        """Estimate the rotation that best transforms a set of vectors to
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: Perhaps: "Estimate the rotation to match two set of vectors".

File: scipy/spatial/transform/rotation.py [hunk 10]
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
...
+            then all values in `weights` are assumed equal to +1.
+            Default is None.
+
+        Returns
+        -------
+        estimated_rotation : `Rotation` instance
+            Contains a single rotation which is the best estimate of the
+            rotation that transforms `vectors` to `transformed_vectors`.
+        sensitivity_matrix : `numpy.ndarray`, shape (3, 3)
+            Scaled covariance matrix of the three component error vector
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: This definitely needs to be explained better. Scaled in what way, how a user should compute a true covariance. Perhaps we should return the true covariance?

File: scipy/spatial/transform/rotation.py [hunk 11]
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
...
         """
         return self.__class__(self._quat[indexer], normalized=True)
+
+    @classmethod
+    def match_vectors(cls, vectors, transformed_vectors, weights=None):
+        # TODO: Better description of coeff?
+        """Estimate the rotation that best transforms a set of vectors to
+        another.
+
+        Given `vectors` expressed in an initial frame A, corresponding
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: I still think that `transformed_vectors` is not a great name.

I don't know, maybe use something neutral like `a` and `b`: ``Find a rotation between frames A and B which best matches a set of unit vectors `a` and `b` observed in these frames. The following optimization problem is solved for the direction cosine matrix `C`: ... ``

My preference is `a = C b`, i.e. C projects from B to A or rotates frame A into B.

Not sure though.

File: scipy/spatial/transform/rotation.py [hunk 12]
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
...
+        vectors : array_like, shape (N, 3)
+            Vectors expressed in initial frame A.
+        transformed_vectors : array_like, shape (N, 3)
+            Vectors expressed in another frame B. Result of applying an
+            arbitrary rotation to `vectors`. Shape must match `vectors`.
+        weights : None or array_like, 1D
+            Coefficients describing the relative importance of the vectors in
+            `vectors`. Number of values in `weights` must match number of
+            vectors specified in `vectors` and `transformed_vectors`. If None,
+            then all values in `weights` are assumed equal to +1.
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: I think "assumed to be equal" is enough.

File: scipy/spatial/transform/rotation.py [hunk 13]
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
...
+            Default is None.
+
+        Returns
+        -------
+        estimated_rotation : `Rotation` instance
+            Contains a single rotation which is the best estimate of the
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: Best estimate of the rotation to match `a` and `b`. Or something like that.

File: scipy/spatial/transform/rotation.py [hunk 14]
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
...
+            if weights.ndim != 1:
+                raise ValueError("Expected `weights` to be 1 dimensional, got "
+                                 "shape {}.".format(weights.shape))
+            if weights.shape[0] != vectors.shape[0]:
+                raise ValueError("Expected `weights` to have number of values "
+                                 "equal to number of input vectors, got "
+                                 "{} values and {} vectors.".format(
+                                    weights.shape[0], vectors.shape[0]))
+
+        B = np.einsum('ji,jk->ik', weights[:, None] * outvecs, vectors)
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: I think this is a slight overuse of `einsum`. There are `np.outer` and other stuff, which might be more efficient. I don't know, perhaps nothing is wrong in using `einsum`, but I somehow don't like it. Judge yourself.

File: scipy/spatial/transform/rotation.py [hunk 15]
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
...
+
+        p = np.diag((k, k, k)) + np.einsum('ij,kj->ik', B, B)
+        try:
+            p = p / psi
+        except ZeroDivisionError:
+            raise ValueError("Three component error vector has infinite "
+                             "covariance. It is impossible to determine the "
+                             "rotation uniquely.")
+
+        return cls.from_dcm(np.einsum('ij,jk->ik', u, vh)), p
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: This is just strange, why do use prefer `einsum` over a simple `dot`?

File: scipy/spatial/transform/rotation.py [hunk 16]
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
...
+                                 "equal to number of input vectors, got "
+                                 "{} values and {} vectors.".format(
+                                    weights.shape[0], vectors.shape[0]))
+
+        B = np.einsum('ji,jk->ik', weights[:, None] * outvecs, vectors)
+        u, s, vh = np.linalg.svd(B)
+        k = s[0]*s[1] + s[1]*s[2] + s[2]*s[0]
+        psi = (s[0]+s[1]) * (s[1]+s[2]) * (s[2]+s[0])
+
+        p = np.diag((k, k, k)) + np.einsum('ij,kj->ik', B, B)
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: `k * np.eye(3)`

And another very strange `einsum`.

File: scipy/spatial/transform/rotation.py [hunk 17]
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
...
+            if weights.shape[0] != vectors.shape[0]:
+                raise ValueError("Expected `weights` to have number of values "
+                                 "equal to number of input vectors, got "
+                                 "{} values and {} vectors.".format(
+                                    weights.shape[0], vectors.shape[0]))
+
+        B = np.einsum('ji,jk->ik', weights[:, None] * outvecs, vectors)
+        u, s, vh = np.linalg.svd(B)
+        k = s[0]*s[1] + s[1]*s[2] + s[2]*s[0]
+        psi = (s[0]+s[1]) * (s[1]+s[2]) * (s[2]+s[0])
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: I believe `zeta` is used in the paper.

And for `k`, I guess `kappa`.

File: scipy/spatial/transform/rotation.py [hunk 18]
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
...
+                raise ValueError("Expected `weights` to be 1 dimensional, got "
+                                 "shape {}.".format(weights.shape))
+            if weights.shape[0] != vectors.shape[0]:
+                raise ValueError("Expected `weights` to have number of values "
+                                 "equal to number of input vectors, got "
+                                 "{} values and {} vectors.".format(
+                                    weights.shape[0], vectors.shape[0]))
+
+        B = np.einsum('ji,jk->ik', weights[:, None] * outvecs, vectors)
+        u, s, vh = np.linalg.svd(B)
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: I suggest to compute `A` right here to make it explicit. Now it is confusing.

File: scipy/spatial/transform/rotation.py [hunk 19]
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
...
+        .. [1] F. Landis Markley,
+                “Attitude determination using vector observations: a fast
+                optimal matrix algorithm,” Journal of Astronautical Sciences,
+                Vol. 41, No.2, 1993, pp. 261-280.
+        """
+        vectors = np.asarray(vectors)
+        if vectors.ndim != 2 or vectors.shape[-1] != 3:
+            raise ValueError("Expected input `vectors` to have shape (N, 3)"
+                             ", got {}.".format(vectors.shape))
+        if vectors.shape[0] == 1:
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: I guess check shape equality here. The check that the number of vectors > 1.

File: scipy/spatial/transform/rotation.py [hunk 20]
```diff
@@ -866,3 +868,106 @@ def __getitem__(self, indexer):
...
+            raise ValueError("Rotation cannot be estimated using a single "
+                             "vector.")
+
+        outvecs = np.asarray(transformed_vectors)
+        if outvecs.ndim != 2 or outvecs.shape[-1] != 3:
```
Comment (nmayorov) [inline scipy/spatial/transform/rotation.py]: Doesn't seem necessary.

**Review by nmayorov — COMMENTED** [review]
Reviewed the latest changes.

**Comment by nmayorov** [comment]
> I need some help with the testing part here. I know I can create test cases by starting with fixed vectors and a given rotation, applying the rotation and adding [gaussian?] noise to get transformed_vectors. How do I ensure that the estimated rotation is within the error bounds? For example, if we were returning the full covariance matrix, I could possibly compute the error vector between estimated_rotation.apply(vectors) and transformed_vectors, and check that the result was within the tolerance given by the covariance matrix. Now since our sensitivity_matrix is scaled, I'm not sure how to proceed.

You need to figure out how the covariance is computed, it's probably a harder part of the paper. Note that there is an appendix. It's quite clear that the covariance is well defined and you can compute it from the noise you are using.

I believe you should add a small gaussian noise and then normalize the vectors. By the way, implement this normalization logic inside the function.

File: scipy/spatial/transform/rotation.py (hunk 11 above)
Comment (adbugger) [reply scipy/spatial/transform/rotation.py]: My thought process behind this name was that it is immediately clear which set of vectors the `estimated_rotation` is being applied to.

File: scipy/spatial/transform/rotation.py (hunk 14 above)
Comment (adbugger) [reply scipy/spatial/transform/rotation.py]: I looked at `np.outer` and that performs outer product of 1D vectors. When given an array of vectors like in our application, it performs outer products of every possible pair.
We need to perform outer products of corresponding vectors and sum the resultant matrices. I think this is a succinct way to do it. I may be wrong here; I'll change it if I find a better way.

**Comment by adbugger** [comment]
Made some changes to the function. The next task is to figure out the covariance matrix.

**Comment by adbugger** [comment]
Added a test case based on my understanding of the covariance matrix. Now that we're normalizing the weights to sum to unity; our previous `sensitivity_matrix` equals the full `covariance_matrix`. 

Cannot rebase as of now due to merge conflicts. I will use the editor to fix those conflicts once we have finalised the `rotation` branch.

**Pull request closed without merging**